### PR TITLE
allow base64 data such as JKS secrets to be passed to the secret object

### PR DESCRIPTION
**Pull request by yugui**
This PR is based on @benjigoldberg's #604. Added some commits to resolve review comments on #604.

Here's a copy&pasted description from #604.
> With Terraform 12, binary objects such as Java Keystores (JKS) can only be correctly loaded into terraform with filebase64. Unfortunately, as described in #518 this breaks the ability for this provider to properly create secrets for these kinds of files, since the data is encoded once by Terraform, and a second time by the Kubernetes control plane.
> 
> As suggested in #518, this adds another attribute to the kubernetes secret which allows users to specify base64data. This data will be decoded from Base64 just prior to being sent to Kubernetes. Additionally, this change should be backward compatible as it adds a new field and does not change the functionality of existing secrets created with this provider.

### CODE REVIEW

**Comment by yugui** [comment]
@pdecat

I created another PR which adds some more commits onto #604 because I need to get #518 resolved but #604 seems to be suspended for a reason.
Would you mind taking another look at this PR?

Although there was a discussion about "consistency with k8s APi" vs "backward compatibility of configurations", I conservatively chose backward compatibility in this PR.
IMO it is better for users so that their secrets will not overwritten by accident.  But I'm happy to revise this PR if you prefer keeping the consistency with k8s.

**Comment by yugui** [comment]
For the review comments on #604,

* [r326574707](https://github.com/terraform-providers/terraform-provider-kubernetes/pull/604#discussion_r326574707):
  > When can this happen?
  > 
  > It would probably be a good thing to add a test for this.
  
  AFAIU, Terraform's provider API does not create such a map. So I removed this case.
  I'll recover the case and adds a test for that if my assumption is wrong.

* [r326577311](https://github.com/terraform-providers/terraform-provider-kubernetes/pull/604#discussion_r326577311):
  > Shouldn't binary values that cannot be encoded as strings be stored in the `base64data` field?
  
  Done.

**Comment by pdecat** [comment]
Hi @yugui, would you mind to `git rebase` your branch on master to cleanup the git history and ease the review?

**Comment by pdecat** [comment]
Also, please note that we were considering abandoning the current implementation from #604 and go with a breaking change: https://github.com/terraform-providers/terraform-provider-kubernetes/pull/604#issuecomment-534500600

It is not decided yet.

**Comment by yugui** [comment]
Thank you for your comments, @pdecat 

> Hi @yugui, would you mind to `git rebase` your branch on master to cleanup the git history and ease the review?

done.

> Also, please note that we were considering abandoning the current implementation from #604 and go with a breaking change: [#604 (comment)](https://github.com/terraform-providers/terraform-provider-kubernetes/pull/604#issuecomment-534500600)
> 
> It is not decided yet.

I saw the discussion. I wonder if I can do something to help it get decided.

**Comment by pdecat** [comment]
Hi @yugui,

given the current situation, I'd say let's do this, and merge this PR as-is to unblock people depending on such secrets to upgrade to terraform 0.12.

Last nitpick, I think `base64data` should be renamed to `base64_data`.

@alexsomesan what do you think?

**Comment by yugui** [comment]
> given the current situation, I'd say let's do this, and merge this PR as-is to unblock people depending on such secrets to upgrade to terraform 0.12.
> 
> Last nitpick, I think `base64data` should be renamed to `base64_data`.

Thank you. I have renamed the field to `base64_data`.

File: kubernetes/resource_kubernetes_secret_test.go [hunk 1]
```diff
@@ -165,7 +168,7 @@ func TestAccKubernetesSecret_importBasic(t *testing.T) {
 		CheckDestroy: testAccCheckKubernetesSecretDestroy,
 		Steps: []resource.TestStep{
 			{
-				Config: testAccKubernetesSecretConfig_basic(name),
+				Config: testAccKubernetesSecretConfig_noBase64(name),
```
Comment (pdecat) [inline kubernetes/resource_kubernetes_secret_test.go]: The `TestAccKubernetesSecret_importBasic` should use the `testAccKubernetesSecretConfig_basic` configuration.

Comment (yugui) [reply kubernetes/resource_kubernetes_secret_test.go]: thanks. done.

**Comment by seboudry** [comment]
Hi!
Any update on when this PR will be merged?

Moreover, to follow naming of `local_file#content_base64` it will be great that the new attribute be named `data_base64` (instead of `base64_data`).
https://www.terraform.io/docs/providers/local/r/file.html#content_base64

**Comment by pascalwhoop** [comment]
the code looks good and this is a blocking issue for us. What is keeping this from being merged? I'm happy to contribute if it's just renaming things.

**Comment by brunocous** [comment]
Hi, how can we help to get this merged?

**Comment by yugui** [comment]
@pdecat Sorry for my late reply.

I'd like to propose an incompatible design change:
* Stop adding `base64_data` argument
* Makes `data` argument always take base64-encoded data as the corresponding Kubernetes API does.

This was the original design which @e96wic proposed.

I am saying this because the original design with `base64_data` cannot work fine with `terraform import`, if I understand correctly.

# An issue with import

I tried to make the integration test `TestAccKubernetesSecret_importBasic` pass.
But I couldn't find the right way to let the import operation reproduce the original configuration.

```
--- FAIL: TestAccKubernetesSecret_importBasic (0.46s)
    testing.go:640: Step 1 error: ImportStateVerify attributes not equivalent. Difference is shown below. Top is actual, bottom is expected.
        
        (map[string]string) (len=2) {
         (string) (len=6) "data.%": (string) (len=1) "3",
         (string) (len=10) "data.three": (string) (len=5) "third"
        }
        
        
        (map[string]string) (len=3) {
         (string) (len=13) "base64_data.%": (string) (len=1) "1",
         (string) (len=17) "base64_data.three": (string) (len=8) "dGhpcmQ=",
         (string) (len=6) "data.%": (string) (len=1) "2"
        }
```

This happens because the import operation does not know a given key is coming from `base64_data` or `data`. This can happen not only in acceptance tests but also in users' environments when they try to import a secret with a binary data.

So, I am sorry, but [my suggestion in #518](https://github.com/terraform-providers/terraform-provider-kubernetes/issues/518#issuecomment-525199127) seems to have been wrong.
I'd like to propose going back to the original design in #518  

What do you think?

**Comment by seboudry** [comment]
Hi!
I don't think making `data` always taking base64-encoded data a good choice.
- data can also be text (this is the common usage)
- it's a breaking change

The `local_file` resources purpose multiple ways to provide data (text, sensitive, base64) and they are mutually exclusive.
https://www.terraform.io/docs/providers/local/r/file.html#content_base64
As this resource is a core Terraform one, I suggest that we should use the same behavior.
Moreover, do you agree to change argument to `data_base64` to follow the same naming?

**Comment by AbdulRahmanAlHamali** [comment]
Thank you for this! Any chance that this would be merged soon?

**Comment by colin-lyman** [comment]
@yugui @benjigoldberg can you sign the CLA so this PR can be merged, for the sake fo the rest of us :)

**Comment by dak1n1** [comment]
Thanks for your work on this PR! We haven't reviewed it yet, but I did look through related issues during our triage process, and opened this new issue to try and address the needs of multiple users. https://github.com/hashicorp/terraform-provider-kubernetes/issues/901 

The goal is to discuss potential approaches and see how the community feels about it, so we can be sure to choose the solution that benefits the most users.

**Comment by TheImplementer** [comment]
Any updates on this?

**Comment by dak1n1** [comment]
Closing since an equivalent feature was released with v2.0.1. https://github.com/hashicorp/terraform-provider-kubernetes/pull/1228

**Pull request closed without merging**